// @ngscaffolding/core/esm2015/lib/services/appAudit/appAudit.service.js
import { Injectable } from '@angular/core';
import { timeout, retry } from 'rxjs/operators';
import { AppSettings, ZuluDateHelper, } from '@ngscaffolding/models';
import * as i0 from "@angular/core";
import * as i1 from "../appSettings/appSettings.query";
import * as i2 from "../userAuthentication/userAuthentication.query";
import * as i3 from "@angular/common/http";
export class AppAuditService {
    constructor(appSettingsQuery, userQuery, http) {
        this.appSettingsQuery = appSettingsQuery;
        this.userQuery = userQuery;
        this.http = http;
    }
    RecordLog(appLog) {
        let apiHome = this.appSettingsQuery.getEntity(AppSettings.apiHome).value;
        if (!appLog.logDate) {
            appLog.logDate = ZuluDateHelper.setGMTDate(new Date());
        }
        if (!appLog.userID) {
            appLog.userID = this.userQuery.getUserId();
        }
        try {
            // This post is a fire and forget. Don't have to authorise either
            this.http
                .post(`${apiHome}/api/v1/applicationLog`, appLog)
                .pipe(timeout(30000), retry(3))
                .subscribe((data) => { }, (err) => {
                console.log('Unable to send AppLog, offline?');
            });
        }
        catch (err) {
            console.log('Unable to send AppLog, offline?');
        }
    }
}
AppAuditService.ɵfac = function AppAuditService_Factory(t) { return new (t || AppAuditService)(i0.ɵɵinject(i1.AppSettingsQuery), i0.ɵɵinject(i2.UserAuthenticationQuery), i0.ɵɵinject(i3.HttpClient)); };
AppAuditService.ɵprov = i0.ɵɵdefineInjectable({ token: AppAuditService, factory: AppAuditService.ɵfac, providedIn: 'root' });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(AppAuditService, [{
        type: Injectable,
        args: [{
                providedIn: 'root',
            }]
    }], function () { return [{ type: i1.AppSettingsQuery }, { type: i2.UserAuthenticationQuery }, { type: i3.HttpClient }]; }, null); })();
