// @ngscaffolding/core/esm2015/lib/services/logging/logging.service.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../notification/notification.service";
import * as i2 from "../appSettings/appSettings.query";
import * as i3 from "../userAuthentication/userAuthentication.query";
import * as i4 from "@angular/common/http";
export class LoggingService {
    constructor(notification, appSettingsQuery, userQuery, http) {
        this.notification = notification;
        this.appSettingsQuery = appSettingsQuery;
        this.userQuery = userQuery;
        this.http = http;
    }
    error(err, methodName = '', showToast = false) {
        if (!methodName) {
            console.error(`Method ${methodName}: Error ${err}`);
        }
        else {
            console.error(`Error ${err}`);
        }
        // TODO: Resume showing toast
        if (showToast) {
            this.notification.showMessage({
                severity: 'error',
                summary: 'Error',
                detail: err.message
            });
        }
    }
    warning(message, methodName = '') {
        if (!methodName) {
            console.warn(`Method ${methodName}: Warning ${message}`);
        }
        else {
            console.warn(`Warning ${message}`);
        }
    }
    info(message, methodName = '', objectInfo = null) {
        if (!methodName) {
            // tslint:disable-next-line:no-console
            console.info(`Info : ${message}`, objectInfo);
        }
        else {
            // tslint:disable-next-line:no-console
            console.info(`[${methodName}] : ${message}`, objectInfo);
        }
    }
}
LoggingService.ɵfac = function LoggingService_Factory(t) { return new (t || LoggingService)(i0.ɵɵinject(i1.NotificationService), i0.ɵɵinject(i2.AppSettingsQuery), i0.ɵɵinject(i3.UserAuthenticationQuery), i0.ɵɵinject(i4.HttpClient)); };
LoggingService.ɵprov = i0.ɵɵdefineInjectable({ token: LoggingService, factory: LoggingService.ɵfac, providedIn: 'root' });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(LoggingService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: i1.NotificationService }, { type: i2.AppSettingsQuery }, { type: i3.UserAuthenticationQuery }, { type: i4.HttpClient }]; }, null); })();
