// @ngscaffolding/core/esm2015/lib/routeGuards/authoriseRoleGuard.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../services/userAuthentication/UserAuthenticationBase";
import * as i2 from "@angular/router";
import * as i3 from "../services/rolesService/roles.service";
export class AuthoriseRoleGuard {
    constructor(authService, router, rolesService) {
        this.authService = authService;
        this.router = router;
        this.rolesService = rolesService;
    }
    canActivate(route, state) {
        if (this.authService.isAuthenticated()) {
            return true;
        }
        // No authority, bye bye.
        this.authService.forceLogon(state.url);
        return false;
    }
}
AuthoriseRoleGuard.ɵfac = function AuthoriseRoleGuard_Factory(t) { return new (t || AuthoriseRoleGuard)(i0.ɵɵinject(i1.UserAuthenticationBase), i0.ɵɵinject(i2.Router), i0.ɵɵinject(i3.RolesService)); };
AuthoriseRoleGuard.ɵprov = i0.ɵɵdefineInjectable({ token: AuthoriseRoleGuard, factory: AuthoriseRoleGuard.ɵfac });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(AuthoriseRoleGuard, [{
        type: Injectable
    }], function () { return [{ type: i1.UserAuthenticationBase }, { type: i2.Router }, { type: i3.RolesService }]; }, null); })();
