// @ngscaffolding/core/esm2015/lib/directives/fill-height.directive.js
import { HostListener, Directive, Input, } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../services/logging/logging.service";
export class FillHeightDirective {
    constructor(el, logger) {
        this.el = el;
        this.logger = logger;
        this.footerElement = null;
        this.fixedHeight = 0;
        this.relativeToParentPercent = 0;
    }
    ngAfterViewInit() {
        this.calculateAndSetElementHeight();
    }
    onResize(event) {
        this.calculateAndSetElementHeight();
    }
    calculateAndSetElementHeight() {
        if (this.relativeToParentPercent > 0) {
            const parentHeight = this.el.nativeElement.parentNode.parentNode
                .parentNode.parentNode.parentNode.parentNode.offsetHeight;
            if (parentHeight > 0) {
                this.el.nativeElement.style.height = `${parentHeight * (this.relativeToParentPercent / 100)}px`;
            }
        }
        else if (this.fixedHeight > 0) {
            this.el.nativeElement.style.height = `${this.fixedHeight}px`;
        }
        else {
            this.el.nativeElement.style.overflow = 'auto';
            let windowHeight = window.innerHeight;
            if (windowHeight === 0) {
                this.logger.error('FillHeight: Window reported zero height');
                windowHeight = 800;
            }
            const elementOffsetTop = this.getElementOffsetTop();
            const elementMarginBottom = this.el.nativeElement.style.marginBottom;
            const footerElementMargin = this.getfooterElementMargin();
            this.el.nativeElement.style.height = `${windowHeight - footerElementMargin - elementOffsetTop - 16}px`;
        }
    }
    getElementOffsetTop() {
        return this.el.nativeElement.getBoundingClientRect().top;
    }
    getfooterElementMargin() {
        if (!this.footerElement) {
            return 0;
        }
        const footerStyle = window.getComputedStyle(this.footerElement);
        return parseInt(footerStyle.height, 10);
    }
}
FillHeightDirective.ɵfac = function FillHeightDirective_Factory(t) { return new (t || FillHeightDirective)(i0.ɵɵdirectiveInject(i0.ElementRef), i0.ɵɵdirectiveInject(i1.LoggingService)); };
FillHeightDirective.ɵdir = i0.ɵɵdefineDirective({ type: FillHeightDirective, selectors: [["", "ngsFillHeight", ""]], hostBindings: function FillHeightDirective_HostBindings(rf, ctx) { if (rf & 1) {
        i0.ɵɵlistener("resize", function FillHeightDirective_resize_HostBindingHandler($event) { return ctx.onResize($event); }, false, i0.ɵɵresolveWindow);
    } }, inputs: { footerElement: "footerElement", fixedHeight: "fixedHeight", relativeToParentPercent: "relativeToParentPercent" } });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(FillHeightDirective, [{
        type: Directive,
        args: [{ selector: '[ngsFillHeight]' }]
    }], function () { return [{ type: i0.ElementRef }, { type: i1.LoggingService }]; }, { footerElement: [{
            type: Input
        }], fixedHeight: [{
            type: Input
        }], relativeToParentPercent: [{
            type: Input
        }], onResize: [{
            type: HostListener,
            args: ['window:resize', ['$event']]
        }] }); })();
