// @ngscaffolding/inputbuilder/esm2015/lib/components/editableTitle/editableTitle.component.js
import { Component, Input, Output, EventEmitter, ViewChild } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common";
import * as i2 from "primeng/button";
import * as i3 from "@angular/forms";
import * as i4 from "primeng/inputtext";
const _c0 = ["inputTitle"];
function EditableTitleComponent_span_0_Template(rf, ctx) { if (rf & 1) {
    const _r4 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "span", 3);
    i0.ɵɵlistener("click", function EditableTitleComponent_span_0_Template_span_click_0_listener() { i0.ɵɵrestoreView(_r4); const ctx_r3 = i0.ɵɵnextContext(); return ctx_r3.mouseClicked(); })("mouseleave", function EditableTitleComponent_span_0_Template_span_mouseleave_0_listener() { i0.ɵɵrestoreView(_r4); const ctx_r5 = i0.ɵɵnextContext(); return ctx_r5.mouseLeave(); })("mouseenter", function EditableTitleComponent_span_0_Template_span_mouseenter_0_listener() { i0.ɵɵrestoreView(_r4); const ctx_r6 = i0.ɵɵnextContext(); return ctx_r6.mouseOver(); });
    i0.ɵɵtext(1);
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r0 = i0.ɵɵnextContext();
    i0.ɵɵadvance(1);
    i0.ɵɵtextInterpolate(ctx_r0.title);
} }
function EditableTitleComponent_button_1_Template(rf, ctx) { if (rf & 1) {
    i0.ɵɵelement(0, "button", 4);
} }
function EditableTitleComponent_input_2_Template(rf, ctx) { if (rf & 1) {
    const _r9 = i0.ɵɵgetCurrentView();
    i0.ɵɵelementStart(0, "input", 5, 6);
    i0.ɵɵlistener("blur", function EditableTitleComponent_input_2_Template_input_blur_0_listener() { i0.ɵɵrestoreView(_r9); const ctx_r8 = i0.ɵɵnextContext(); return ctx_r8.isEditing = false; })("ngModelChange", function EditableTitleComponent_input_2_Template_input_ngModelChange_0_listener($event) { i0.ɵɵrestoreView(_r9); const ctx_r10 = i0.ɵɵnextContext(); return ctx_r10.editTitle = $event; })("change", function EditableTitleComponent_input_2_Template_input_change_0_listener() { i0.ɵɵrestoreView(_r9); const ctx_r11 = i0.ɵɵnextContext(); return ctx_r11.textChanged(); });
    i0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r2 = i0.ɵɵnextContext();
    i0.ɵɵproperty("ngModel", ctx_r2.editTitle);
} }
export class EditableTitleComponent {
    constructor() {
        this.titleChanged = new EventEmitter();
        this.displayIcon = false;
        this.isEditing = false;
    }
    ngOnChanges(changes) {
        if (changes.title) {
            this.editTitle = changes.title.currentValue;
            this.applyObject(changes.configObject);
        }
        if (changes.configObject) {
            this.applyObject(changes.configObject);
        }
    }
    // update Title with string replacement @@key## style
    applyObject(config) {
        let newTitle = this.title;
        if (config) {
            for (const key in config) {
                if (config.hasOwnProperty(key)) {
                    newTitle = newTitle.replace(`@@${key}##`, config[key]);
                }
            }
            this.editTitle = newTitle;
        }
    }
    mouseOver() {
        if (!this.readOnly) {
            this.displayIcon = true;
        }
    }
    mouseLeave() {
        this.displayIcon = false;
    }
    mouseClicked() {
        if (!this.readOnly) {
            this.isEditing = true;
            this.displayIcon = false;
        }
        setTimeout(() => {
            // this will make the execution after the above boolean has changed
            this.inputTitleElement.nativeElement.focus();
            this.inputTitleElement.nativeElement.select();
        }, 0);
    }
    textChanged() {
        this.titleChanged.emit(this.editTitle);
    }
}
EditableTitleComponent.ɵfac = function EditableTitleComponent_Factory(t) { return new (t || EditableTitleComponent)(); };
EditableTitleComponent.ɵcmp = i0.ɵɵdefineComponent({ type: EditableTitleComponent, selectors: [["ngs-editable-title"]], viewQuery: function EditableTitleComponent_Query(rf, ctx) { if (rf & 1) {
        i0.ɵɵviewQuery(_c0, true);
    } if (rf & 2) {
        let _t;
        i0.ɵɵqueryRefresh(_t = i0.ɵɵloadQuery()) && (ctx.inputTitleElement = _t.first);
    } }, inputs: { title: "title", configObject: "configObject", readOnly: "readOnly" }, outputs: { titleChanged: "titleChanged" }, features: [i0.ɵɵNgOnChangesFeature], decls: 3, vars: 3, consts: [["class", "title-display", 3, "click", "mouseleave", "mouseenter", 4, "ngIf"], ["pButton", "", "class", "p-button-rounded p-button-info p-button-text p-mr-2 p-mb-2", "type", "button", "icon", "pi pi-pencil", 4, "ngIf"], ["class", "large-edit", "type", "text", "pInputText", "", 3, "ngModel", "blur", "ngModelChange", "change", 4, "ngIf"], [1, "title-display", 3, "click", "mouseleave", "mouseenter"], ["pButton", "", "type", "button", "icon", "pi pi-pencil", 1, "p-button-rounded", "p-button-info", "p-button-text", "p-mr-2", "p-mb-2"], ["type", "text", "pInputText", "", 1, "large-edit", 3, "ngModel", "blur", "ngModelChange", "change"], ["inputTitle", ""]], template: function EditableTitleComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵtemplate(0, EditableTitleComponent_span_0_Template, 2, 1, "span", 0);
        i0.ɵɵtemplate(1, EditableTitleComponent_button_1_Template, 1, 0, "button", 1);
        i0.ɵɵtemplate(2, EditableTitleComponent_input_2_Template, 2, 1, "input", 2);
    } if (rf & 2) {
        i0.ɵɵproperty("ngIf", !ctx.isEditing);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.displayIcon);
        i0.ɵɵadvance(1);
        i0.ɵɵproperty("ngIf", ctx.isEditing);
    } }, directives: [i1.NgIf, i2.ButtonDirective, i3.DefaultValueAccessor, i4.InputText, i3.NgControlStatus, i3.NgModel], styles: [".toolbarButton[_ngcontent-%COMP%]{border-radius:0x!important;margin-left:2px}.large-edit[_ngcontent-%COMP%]{font-size:20px;width:50%}.tipIcon[_ngcontent-%COMP%]{font-size:12px;margin-left:6px;top:-4px}.title-display[_ngcontent-%COMP%]{margin-left:.5em}"] });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(EditableTitleComponent, [{
        type: Component,
        args: [{
                selector: 'ngs-editable-title',
                templateUrl: 'editableTitle.component.html',
                styleUrls: ['editableTitle.component.scss']
            }]
    }], null, { inputTitleElement: [{
            type: ViewChild,
            args: ['inputTitle', { static: false }]
        }], title: [{
            type: Input
        }], configObject: [{
            type: Input
        }], readOnly: [{
            type: Input
        }], titleChanged: [{
            type: Output
        }] }); })();
