// @ngscaffolding/core/esm2015/lib/services/widgetsService/widget.service.js
import { Injectable } from '@angular/core';
import { combineLatest } from 'rxjs';
import { finalize, take } from 'rxjs/operators';
import { AppSettings } from '@ngscaffolding/models';
import * as i0 from "@angular/core";
import * as i1 from "@angular/common/http";
import * as i2 from "./widget.store";
import * as i3 from "./widget.query";
import * as i4 from "../appSettings/appSettings.query";
import * as i5 from "../userAuthentication/userAuthentication.query";
import * as i6 from "../logging/logging.service";
import * as i7 from "../rolesService/roles.service";
export class WidgetService {
    constructor(http, widgetStore, widgetQuery, appSettingsQuery, authQuery, log, rolesService) {
        this.http = http;
        this.widgetStore = widgetStore;
        this.widgetQuery = widgetQuery;
        this.appSettingsQuery = appSettingsQuery;
        this.authQuery = authQuery;
        this.log = log;
        this.rolesService = rolesService;
        this.className = 'core.WidgetService';
        // First Time load away
        this.widgetStore.setLoading(false);
        // Wait for settings, then load from server
        combineLatest(this.authQuery.authenticated$, this.appSettingsQuery.selectEntity(AppSettings.apiHome)).subscribe(([authenticated, apiHome]) => {
            if (authenticated && apiHome) {
                this.apiHome = apiHome.value;
                this.widgetQuery
                    .selectLoading()
                    .pipe(take(1))
                    .subscribe(loading => {
                    if (!loading) {
                        this.downloadWidgetItems();
                    }
                });
            }
            else if (!authenticated) {
                this.widgetStore.remove();
            }
        });
    }
    downloadWidgetItems() {
        // Mark loading status
        this.widgetStore.setLoading(true);
        this.http
            .get(this.apiHome + '/api/v1/widgets')
            .pipe(finalize(() => {
            this.widgetStore.setLoading(false);
        }))
            .subscribe(widgetItems => {
            this.widgetStore.add(widgetItems);
        });
    }
}
WidgetService.ɵfac = function WidgetService_Factory(t) { return new (t || WidgetService)(i0.ɵɵinject(i1.HttpClient), i0.ɵɵinject(i2.WidgetStore), i0.ɵɵinject(i3.WidgetQuery), i0.ɵɵinject(i4.AppSettingsQuery), i0.ɵɵinject(i5.UserAuthenticationQuery), i0.ɵɵinject(i6.LoggingService), i0.ɵɵinject(i7.RolesService)); };
WidgetService.ɵprov = i0.ɵɵdefineInjectable({ token: WidgetService, factory: WidgetService.ɵfac, providedIn: 'root' });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(WidgetService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: i1.HttpClient }, { type: i2.WidgetStore }, { type: i3.WidgetQuery }, { type: i4.AppSettingsQuery }, { type: i5.UserAuthenticationQuery }, { type: i6.LoggingService }, { type: i7.RolesService }]; }, null); })();
