// @ngscaffolding/core/esm2015/lib/services/coreErrorHandler/coreErrorHandler.service.js
import { ErrorHandler, Injectable } from '@angular/core';
import { ErrorModel, AppSettings } from '@ngscaffolding/models';
import * as i0 from "@angular/core";
import * as i1 from "../appSettings/appSettings.service";
import * as i2 from "@angular/common/http";
export class CoreErrorHandlerService extends ErrorHandler {
    constructor(appSettingsService, http) {
        super();
        this.appSettingsService = appSettingsService;
        this.http = http;
    }
    logError(error, source = null) { }
    processError(error, source = null) { }
    handleError(error, source = null) {
        super.handleError(error);
        if (this.appSettingsService.getValue(AppSettings.errorLogConsole)) {
            console.error(error.message);
        }
        if (this.appSettingsService.getValue(AppSettings.errorLogServer)) {
            const errorModel = new ErrorModel(error);
            if (source) {
                errorModel.source = source;
            }
            if (error.stack) {
                errorModel.stackTrace = error.stack;
            }
            // Consume any errors here. Otherwise we will just get stuck
            try {
                // This post is a fire and forget. Don't have to authorise either
                this.http.post(this.appSettingsService.getValue(AppSettings.apiHome) + '/api/v1/error', errorModel).subscribe(data => {
                    // alert('ID: ' + data.id);
                }, err => {
                    console.log('Unable to send Error to Server, offline?');
                });
            }
            catch (err) {
                console.log('Unable to send Error to Server, offline?');
            }
        }
        if (this.appSettingsService.getValue(AppSettings.errorShowUser)) {
            // TODO: Show User Error
        }
    }
}
CoreErrorHandlerService.ɵfac = function CoreErrorHandlerService_Factory(t) { return new (t || CoreErrorHandlerService)(i0.ɵɵinject(i1.AppSettingsService), i0.ɵɵinject(i2.HttpClient)); };
CoreErrorHandlerService.ɵprov = i0.ɵɵdefineInjectable({ token: CoreErrorHandlerService, factory: CoreErrorHandlerService.ɵfac, providedIn: 'root' });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(CoreErrorHandlerService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: i1.AppSettingsService }, { type: i2.HttpClient }]; }, null); })();
