// @ngscaffolding/core/esm2015/lib/services/componentLoader.service.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class ComponentLoaderService {
    constructor(loader, injector) {
        this.loader = loader;
        this.injector = injector;
        this.componentRegistry = {
            'lazy-component': {
                modulePath: './lazy-components/lazy-components.module#LazyComponentsModule',
                moduleRef: null
            }
        };
    }
    registerComponent(name, modulePath = null) {
        if (!this.componentRegistry[name]) {
            this.componentRegistry[name] = {
                modulePath: modulePath,
                moduleRef: null
            };
        }
    }
    loadComponent(componentTag) {
        const cmpRegistryEntry = this.componentRegistry[componentTag];
        if (!cmpRegistryEntry) {
            throw new Error(`Unrecognized component "${componentTag}". Make sure it is registered in the component registry`);
        }
        if (cmpRegistryEntry.modulePath) {
            // Provided a path so this is a lazy loaded component
            if (cmpRegistryEntry.moduleRef) {
                return new Promise(resolve => {
                    const componentInstance = document.createElement(componentTag);
                    resolve(componentInstance);
                });
            }
            else {
                const path = cmpRegistryEntry.modulePath;
                return new Promise((resolve, reject) => {
                    this.loader
                        .load(path)
                        .then(moduleFactory => {
                        const moduleRef = moduleFactory.create(this.injector).instance;
                        cmpRegistryEntry.moduleRef = moduleRef;
                        // instantiate the component
                        const componentInstance = document.createElement(componentTag);
                        resolve(componentInstance);
                    })
                        .catch(err => {
                        console.error('error loading module', err);
                        reject(err);
                    });
                });
            }
        }
        else {
            // No path so simple Angular Element already webpacked
            return new Promise((resolve, reject) => {
                const componentInstance = document.createElement(componentTag);
                resolve(componentInstance);
            });
        }
    }
}
ComponentLoaderService.ɵfac = function ComponentLoaderService_Factory(t) { return new (t || ComponentLoaderService)(i0.ɵɵinject(i0.NgModuleFactoryLoader), i0.ɵɵinject(i0.Injector)); };
ComponentLoaderService.ɵprov = i0.ɵɵdefineInjectable({ token: ComponentLoaderService, factory: ComponentLoaderService.ɵfac, providedIn: 'root' });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(ComponentLoaderService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: i0.NgModuleFactoryLoader }, { type: i0.Injector }]; }, null); })();
