// @ngscaffolding/core/esm2015/lib/pipes/ngsDateTime.pipe.js
import * as dateFormatimp from 'dateformat';
import { Pipe } from '@angular/core';
import { AppSettings } from '@ngscaffolding/models';
import * as i0 from "@angular/core";
import * as i1 from "../services/appSettings/appSettings.query";
const dateFormat = dateFormatimp;
export class NgsDateTimePipe {
    constructor(appSettings) {
        this.appSettings = appSettings;
    }
    transform(inputDate) {
        if (inputDate) {
            // If a string gets through, convert to date object
            if (typeof inputDate === 'string' || inputDate instanceof String) {
                inputDate = new Date(inputDate);
            }
            const userTimezoneOffset = inputDate.getTimezoneOffset() * 60000;
            const zuluDate = new Date(inputDate.getTime() + userTimezoneOffset);
            const format = this.appSettings.getEntity(AppSettings.dateTimeFormat);
            if (format && format.value) {
                return dateFormat(zuluDate, format.value);
            }
            else {
                return dateFormat(zuluDate, 'default');
            }
        }
        else {
            return '';
        }
    }
}
NgsDateTimePipe.ɵfac = function NgsDateTimePipe_Factory(t) { return new (t || NgsDateTimePipe)(i0.ɵɵdirectiveInject(i1.AppSettingsQuery)); };
NgsDateTimePipe.ɵpipe = i0.ɵɵdefinePipe({ name: "ngsDateTime", type: NgsDateTimePipe, pure: true });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(NgsDateTimePipe, [{
        type: Pipe,
        args: [{ name: 'ngsDateTime' }]
    }], function () { return [{ type: i1.AppSettingsQuery }]; }, null); })();
