// @ngscaffolding/core/esm2015/lib/pipes/truncateText.pipe.js
import { Pipe } from '@angular/core';
import * as i0 from "@angular/core";
export class TruncateTextPipe {
    transform(value, limit = 40, trail = '…') {
        let result = value || '';
        if (value) {
            const words = value.split(/\s+/);
            if (words.length > Math.abs(limit)) {
                if (limit < 0) {
                    limit *= -1;
                    result =
                        trail + words.slice(words.length - limit, words.length).join(' ');
                }
                else {
                    result = words.slice(0, limit).join(' ') + trail;
                }
            }
        }
        return result;
    }
}
TruncateTextPipe.ɵfac = function TruncateTextPipe_Factory(t) { return new (t || TruncateTextPipe)(); };
TruncateTextPipe.ɵpipe = i0.ɵɵdefinePipe({ name: "truncateText", type: TruncateTextPipe, pure: true });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(TruncateTextPipe, [{
        type: Pipe,
        args: [{
                name: 'truncateText'
            }]
    }], null, null); })();
