// @ngscaffolding/core/esm2015/lib/services/menu/menu.query.js
import { __decorate } from "tslib";
import { Injectable } from '@angular/core';
import { QueryEntity, QueryConfig, Order } from '@datorama/akita';
import * as i0 from "@angular/core";
import * as i1 from "./menu.store";
let MenuQuery = class MenuQuery extends QueryEntity {
    constructor(store) {
        super(store);
        this.store = store;
        this.folders$ = this.selectAll({
            filterBy: [entity => entity.type === "folder" /* Folder */]
        });
    }
};
MenuQuery.ɵfac = function MenuQuery_Factory(t) { return new (t || MenuQuery)(i0.ɵɵinject(i1.MenuStore)); };
MenuQuery.ɵprov = i0.ɵɵdefineInjectable({ token: MenuQuery, factory: MenuQuery.ɵfac, providedIn: 'root' });
MenuQuery = __decorate([
    QueryConfig({
        sortBy: 'order',
        sortByOrder: Order.ASC // Order.DESC
    })
], MenuQuery);
export { MenuQuery };
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(MenuQuery, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: i1.MenuStore }]; }, null); })();
