// @ngscaffolding/core/esm2015/lib/services/auditLog/auditLog.service.js
import { Injectable } from '@angular/core';
import { AppSettings, ZuluDateHelper } from '@ngscaffolding/models';
import { timeout, retry, finalize } from 'rxjs/operators';
import { v4 as uuid } from 'uuid';
import * as i0 from "@angular/core";
import * as i1 from "./auditLog.store";
import * as i2 from "./auditLog.query";
import * as i3 from "../appSettings/appSettings.query";
import * as i4 from "../userAuthentication/userAuthentication.query";
import * as i5 from "@angular/common/http";
export class AuditLogService {
    constructor(auditLogStore, auditLogQuery, appSettingsQuery, userQuery, http) {
        this.auditLogStore = auditLogStore;
        this.auditLogQuery = auditLogQuery;
        this.appSettingsQuery = appSettingsQuery;
        this.userQuery = userQuery;
        this.http = http;
        this.polling = 30000;
        this.retryVal = 3;
        this.isSending = false;
        this.defaultLog = {};
        appSettingsQuery
            .selectEntity(AppSettings.mobileDefaultPolling)
            .subscribe(val => (this.polling = val.value > 0 ? val.value : 30000));
        appSettingsQuery
            .selectEntity(AppSettings.mobileDefaultRetries)
            .subscribe(val => (this.retryVal = val.value > 0 ? val.value : 3));
        this.StartPolling();
    }
    SetDefault(defaultLog) {
        this.defaultLog = defaultLog;
    }
    RecordLog(auditLog) {
        const workingLog = Object.assign(Object.assign({}, this.defaultLog), auditLog);
        workingLog.id = uuid();
        if (!workingLog.logDate) {
            workingLog.logDate = ZuluDateHelper.setGMTDate(new Date());
        }
        if (!workingLog.userID) {
            workingLog.userID = this.userQuery.getUserId();
        }
        this.auditLogStore.add(workingLog);
        try {
        }
        catch (err) {
            console.log('Unable to send AppLog, offline?');
        }
    }
    StartPolling() {
        setInterval(_ => {
            if (!this.isSending) {
                this.isSending = true;
                this.SendLogEntries();
            }
        }, this.polling);
    }
    SendLogEntries() {
        const apiHome = this.appSettingsQuery.getEntity(AppSettings.apiHome).value;
        const logEntries = this.auditLogQuery.getAll();
        if (logEntries && logEntries.length > 0) {
            const keys = logEntries.map(log => log.id);
            // This post is a fire and forget. Don't have to authorise either
            this.http
                .post(`${apiHome}/api/v1/auditlog`, logEntries)
                .pipe(timeout(30000), retry(3), finalize(() => {
                this.isSending = false;
            }))
                .subscribe(data => {
                this.auditLogStore.remove(keys);
                // keys.forEach(key => this.auditLogStore.remove(({ id }) => id === key));
            }, err => {
                console.log('Unable to send AppLog, offline?');
            });
        }
        else {
            this.isSending = false;
        }
    }
}
AuditLogService.ɵfac = function AuditLogService_Factory(t) { return new (t || AuditLogService)(i0.ɵɵinject(i1.AuditLogStore), i0.ɵɵinject(i2.AuditLogQuery), i0.ɵɵinject(i3.AppSettingsQuery), i0.ɵɵinject(i4.UserAuthenticationQuery), i0.ɵɵinject(i5.HttpClient)); };
AuditLogService.ɵprov = i0.ɵɵdefineInjectable({ token: AuditLogService, factory: AuditLogService.ɵfac, providedIn: 'root' });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(AuditLogService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: i1.AuditLogStore }, { type: i2.AuditLogQuery }, { type: i3.AppSettingsQuery }, { type: i4.UserAuthenticationQuery }, { type: i5.HttpClient }]; }, null); })();
