// @ngscaffolding/core/esm2015/lib/services/userAuthentication/userAuthentication.query.js
import { Injectable } from '@angular/core';
import { Query, toBoolean } from '@datorama/akita';
import * as i0 from "@angular/core";
import * as i1 from "./userAuthentication.store";
export class UserAuthenticationQuery extends Query {
    constructor(store) {
        super(store);
        this.store = store;
        this.authenticated$ = this.select(state => state.authenticated);
        this.currentUser$ = this.select(state => state.userDetails);
    }
    isAuthenticated() {
        return toBoolean(this.getValue().authenticated);
    }
    getUserId() {
        if (this.isAuthenticated() && this.getValue().userDetails) {
            return this.getValue().userDetails.userId;
        }
        else {
            return null;
        }
    }
    getUser() {
        return this.getValue().userDetails;
    }
}
UserAuthenticationQuery.ɵfac = function UserAuthenticationQuery_Factory(t) { return new (t || UserAuthenticationQuery)(i0.ɵɵinject(i1.AuthenticationStore)); };
UserAuthenticationQuery.ɵprov = i0.ɵɵdefineInjectable({ token: UserAuthenticationQuery, factory: UserAuthenticationQuery.ɵfac, providedIn: 'root' });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(UserAuthenticationQuery, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: i1.AuthenticationStore }]; }, null); })();
