// @ngscaffolding/core/esm2015/lib/core.module.js
import { NgModule } from '@angular/core';
import { HttpClientModule } from '@angular/common/http';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
// Pipes
import { ButtonColourPipe } from './pipes/buttonColour.pipe';
import { NgsDateTimePipe } from './pipes/ngsDateTime.pipe';
import { NgsDatePipe } from './pipes/ngsDate.pipe';
import { TruncateTextPipe } from './pipes/truncateText.pipe';
// Directives
import { FillHeightDirective } from './directives/fill-height.directive';
// Components
import { DialogWindowComponent } from './components/dialogWindow/dialogWindow.component';
import { VERSION } from './version';
import * as i0 from "@angular/core";
import * as i1 from "./services/versions/versions.service";
export class CoreModule {
    constructor(versions) {
        versions.addVersion('@ngscaffolding/core', VERSION.version);
    }
    static forRoot() {
        return {
            ngModule: CoreModule
        };
    }
}
CoreModule.ɵmod = i0.ɵɵdefineNgModule({ type: CoreModule });
CoreModule.ɵinj = i0.ɵɵdefineInjector({ factory: function CoreModule_Factory(t) { return new (t || CoreModule)(i0.ɵɵinject(i1.VersionsService)); }, imports: [[CommonModule, FormsModule, HttpClientModule]] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(CoreModule, { declarations: [FillHeightDirective,
        ButtonColourPipe,
        NgsDatePipe,
        NgsDateTimePipe,
        TruncateTextPipe,
        DialogWindowComponent], imports: [CommonModule, FormsModule, HttpClientModule], exports: [ButtonColourPipe,
        NgsDatePipe,
        NgsDateTimePipe,
        TruncateTextPipe,
        FillHeightDirective,
        DialogWindowComponent] }); })();
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(CoreModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule, FormsModule, HttpClientModule],
                declarations: [
                    FillHeightDirective,
                    ButtonColourPipe,
                    NgsDatePipe,
                    NgsDateTimePipe,
                    TruncateTextPipe,
                    DialogWindowComponent
                ],
                exports: [
                    ButtonColourPipe,
                    NgsDatePipe,
                    NgsDateTimePipe,
                    TruncateTextPipe,
                    FillHeightDirective,
                    DialogWindowComponent
                ]
            }]
    }], function () { return [{ type: i1.VersionsService }]; }, null); })();
