// @ngscaffolding/core/esm2015/lib/components/dialogWindow/dialogWindow.component.js
import { Component, EventEmitter, Output } from '@angular/core';
import * as i0 from "@angular/core";
export class DialogWindowComponent {
    constructor() {
        this.closed = new EventEmitter();
    }
    ngOnChanges(changes) { }
    ngOnDestroy() { }
    ngOnInit() { }
    cancel(event) {
        this.closed.emit(null);
    }
    save(event) {
    }
}
DialogWindowComponent.ɵfac = function DialogWindowComponent_Factory(t) { return new (t || DialogWindowComponent)(); };
DialogWindowComponent.ɵcmp = i0.ɵɵdefineComponent({ type: DialogWindowComponent, selectors: [["ngs-dialog-window"]], outputs: { closed: "closed" }, features: [i0.ɵɵNgOnChangesFeature], decls: 2, vars: 0, template: function DialogWindowComponent_Template(rf, ctx) { if (rf & 1) {
        i0.ɵɵelementStart(0, "h1");
        i0.ɵɵtext(1, "Dialog Window Base Class");
        i0.ɵɵelementEnd();
    } }, styles: [""] });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(DialogWindowComponent, [{
        type: Component,
        args: [{
                selector: 'ngs-dialog-window',
                templateUrl: 'dialogWindow.component.html',
                styleUrls: ['dialogWindow.component.scss']
            }]
    }], null, { closed: [{
            type: Output
        }] }); })();
