// @ngscaffolding/core/esm2015/lib/services/versions/versions.service.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "../logging/logging.service";
export class SoftwareVersion {
}
export class VersionsService {
    constructor(logging) {
        this.logging = logging;
        this.versions = [];
    }
    addVersion(module, version, isAppModule = false) {
        this.logging.info(`Loading Module ${module} Version ${version} IsAppModule ${isAppModule}`);
        this.versions.push({
            module: module,
            version: version,
            isAppModule: isAppModule
        });
    }
    getVersions() {
        return this.versions;
    }
    getVersion(moduleName) {
        const foundModule = this.versions.find(v => v.module === moduleName);
        if (foundModule) {
            return foundModule.version;
        }
        else {
            return null;
        }
    }
}
VersionsService.ɵfac = function VersionsService_Factory(t) { return new (t || VersionsService)(i0.ɵɵinject(i1.LoggingService)); };
VersionsService.ɵprov = i0.ɵɵdefineInjectable({ token: VersionsService, factory: VersionsService.ɵfac, providedIn: 'root' });
/*@__PURE__*/ (function () { i0.ɵsetClassMetadata(VersionsService, [{
        type: Injectable,
        args: [{
                providedIn: 'root'
            }]
    }], function () { return [{ type: i1.LoggingService }]; }, null); })();
